es5-shim.js

```diff
@@ -1921,20 +1921,93 @@
+                var nextSecond = source[+charIndex + 2] || null;
+
+                if (char === '(') {
+                    if (prevFirst && prevFirst === '\\') {
+                        continue;
+                    }
+
+                    if (
+                        nextFirst && nextFirst === '?' && nextSecond && nextSecond === ':'
+                        || nextFirst && nextFirst === '?' && nextSecond && nextSecond === '='
+                        || nextFirst && nextFirst === '?' && nextSecond && nextSecond === '!'
+                    ) {
+                        groupNumberCounter += 1;
+                        result.push({
+                            number: groupNumberCounter,
+                            mayBeMissing: true,
+                            isResulting: false
+                        });
+                        continue;
+                    }
+
+                    result.push({
+                        number: ++groupNumberCounter,
+                        mayBeMissing: false,
+                        isResulting: true
+                    });
+                }
+
+                if (char === ')') {
+                    var groupIndex = groupNumberCounter - 1;
+
+                    if (prevFirst && prevFirst === '\\') {
+                        continue;
+                    }
+
+                    if (result[groupIndex] && !result[groupIndex].isResulting) {
+                        result.splice(groupIndex, 1);
+                        groupNumberCounter = --groupNumberCounter;
+                        continue;
+                    }
+
+                    if (nextFirst && (nextFirst === '?' || nextFirst === '*')) {
+                        result[groupIndex].mayBeMissing = true;
+                    }
+                }
+            }
+
+            return result;
+        };
+
         StringPrototype.replace = function replace(searchValue, replaceValue) {
             var isFn = isCallable(replaceValue);
             var hasCapturingGroups = isRegex(searchValue) && (/\)[*?]/).test(searchValue.source);
+            var extendedGroups = getExtendedGroups(searchValue);
+
             if (!isFn || !hasCapturingGroups) {
                 return str_replace.call(this, searchValue, replaceValue);
             } else {
                 var wrappedReplaceValue = function (match) {
-                    var length = arguments.length;
-                    var originalLastIndex = searchValue.lastIndex;
-                    searchValue.lastIndex = 0; // eslint-disable-line no-param-reassign
-                    var args = searchValue.exec(match) || [];
-                    searchValue.lastIndex = originalLastIndex; // eslint-disable-line no-param-reassign
-                    pushCall(args, arguments[length - 2], arguments[length - 1]);
-                    return replaceValue.apply(this, args);
+                    var groups = ArrayPrototype.slice.call(arguments, 1, -2);
```

Comment:
`delete Array.prototype.slice` would break this code.
```suggestion
                    var groups = arraySlice(arguments, 1, -2);
```